File: src/year2019/day23/day23.rs
//! [Day 23: Category Six](https://adventofcode.com/2019/day/23)

use std::collections::{HashSet, VecDeque};

use intcode::{Computer, State};

enum SchedState {
    Idle,
    Active,
}

struct Message {
    dest_id: i64,
    x: i64,
    y: i64,
}

struct SchedResult {
    messages: VecDeque<Message>,
    state: SchedState,
}

impl SchedResult {
    const fn new() -> Self {
        Self {
            messages: VecDeque::new(),
            state: SchedState::Active,
        }
    }
    fn push(&mut self, dest_id: i64, x: i64, y: i64) {
        self.messages.push_back(Message { dest_id, x, y });
    }
}

trait Task {
    fn sched(&mut self, idle_count: usize) -> SchedResult;
    fn receive(&mut self, x: i64, y: i64);
}

struct Network {
    nodes: Vec<Node>,
    nat: Nat,
    idle: HashSet<usize>,
}

impl Network {
    fn new(program: &str) -> Self {
        let mut nodes = Vec::new();
        let computer = Computer::load(program);
        for id in 0..50 {
            nodes.push(Node::new(id, &computer));
        }

        Self {
            nodes,
            nat: Nat::new(),
            idle: HashSet::new(),
        }
    }

    fn run(&mut self) -> (i64, i64) {
        let mut part1 = 0;
        let mut idle_y = HashSet::new();

        let mut id = 0;

        loop {
            let idle_count = self.idle.len();
            let a = if id == 255 {
                self.nat.sched(idle_count)
            } else {
                self.nodes[id].sched(idle_count)
            };

            match a.state {
                SchedState::Active => {
                    self.idle.remove(&id);
                }
                SchedState::Idle => {
                    self.idle.insert(id);
                }
            };

            for m in a.messages {
                // deal with part 2
                if id == 255 && idle_count == 50 && !idle_y.insert(m.y) {
                    return (part1, m.y);
                }

                if m.dest_id == 255 {
                    self.nat.receive(m.x, m.y);

                    // deal with part 1
                    if part1 == 0 {
                        part1 = m.y;
                    }
                } else {
                    let dest_idx = usize::try_from(m.dest_id).unwrap();
                    self.nodes[dest_idx].receive(m.x, m.y);
                }
            }

            id = match id {
                0..=48 => id + 1,
                49 => 255,
                255 => 0,
                _ => unreachable!(),
            };
        }
    }
}

struct Nat {
    last_x: i64,
    last_y: i64,
    input: VecDeque<i64>,
}
impl Nat {
    const fn new() -> Self {
        Self {
            last_x: 0,
            last_y: 0,
            input: VecDeque::new(),
        }
    }
}

impl Task for Nat {
    fn sched(&mut self, idle_count: usize) -> SchedResult {
        while self.input.len() >= 2 {
            self.last_x = self.input.pop_front().unwrap();
            self.last_y = self.input.pop_front().unwrap();
        }

        let mut result = SchedResult::new();

        if idle_count == 50 {
            // if !self.idle_y.insert(self.last_y) {
            //     result.state = SchedState::Halt;
            // } else {
            result.push(0, self.last_x, self.last_y);
            // }
        }

        result
    }

    fn receive(&mut self, x: i64, y: i64) {
        self.input.push_back(x);
        self.input.push_back(y);
    }
}

struct Node {
    id: i64,
    computer: Computer,
    output: VecDeque<i64>,
}

impl Node {
    fn new(id: i64, computer: &Computer) -> Self {
        let mut computer = computer.clone();
        computer.push(id);
        let state = computer.run();
        assert_eq!(state, State::Input);

        Self {
            id,
            computer: computer.clone(),
            output: VecDeque::new(),
        }
    }
}

impl Task for Node {
    fn sched(&mut self, _idle_count: usize) -> SchedResult {
        let nb_recv = if self.computer.input_len() == 0 {
            self.computer.push(-1);
            0
        } else {
            self.computer.input_len() / 3
        };

        while let State::Output(num) = self.computer.run() {
            self.output.push_back(num);
        }

        let mut result = SchedResult::new();

        while self.output.len() >= 3 {
            result.push(
                self.output.pop_front().unwrap(),
                self.output.pop_front().unwrap(),
                self.output.pop_front().unwrap(),
            );
        }

        result.state = if nb_recv == 0 {
            SchedState::Idle
        } else {
            SchedState::Active
        };

        result
    }

    fn receive(&mut self, x: i64, y: i64) {
        self.computer.push(self.id);
        self.computer.push(0);
        self.computer.push(x);
        self.computer.push(y);
    }
}

/// # Panics
/// over malformed input
#[must_use]
pub fn solve(data: &str) -> (i64, i64) {
    Network::new(data).run()
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

File: src/year2019/day23/day23.py
#!/usr/bin/env python3
# [Day 23: Category Six](https://adventofcode.com/2019/day/23)

import atexit
import itertools
import sys
import time
from pathlib import Path

filename = sys.argv[1] if len(sys.argv) > 1 else "input.txt"
software = Path(filename).read_text()
if "--elapsed" in sys.argv:
    sys.argv.remove("--elapsed")
    start_time_ns = time.time_ns()
    atexit.register(lambda: print(f"elapsed: {(time.time_ns() - start_time_ns) / 1_000_000}ms"))


sys.path.append(Path(__file__).parent.parent.as_posix())
from intcode.Intcode import Computer  # noqa


class Node:
    network = {}
    idle = set()

    def __init__(self, id):
        self.network[id] = self
        self.id = id

        self.computer = Computer()

        # NAT is not a true Intcode computer (software segfault...)
        if id == 255:
            self.last_X = 0  # the last NAT message received
            self.last_Y = 0
            self.idle_Y = set()
        else:
            self.computer.load(software)
            self.computer.input.append(id)
            state = self.computer.run()
            assert state == "read" and len(self.computer.output) == 0

    def sched(self, part):
        answer = None

        if self.id == 255:
            while len(self.computer.input) > 0:
                _ = self.computer.input.popleft()  # seems to be always 24
                _ = self.computer.input.popleft()  # weird, see below
                X = self.computer.input.popleft()
                Y = self.computer.input.popleft()

                self.last_X = X
                self.last_Y = Y

            # if ell nodes are idle
            if len(self.idle) == 50:
                self.send_to(0, self.last_X, self.last_Y)

                if part == 2:
                    if self.last_Y in self.idle_Y:
                        return self.last_Y
                    self.idle_Y.add(self.last_Y)

            return

        if len(self.computer.input) == 0:
            self.computer.input.append(-1)
            nb_recv = 0
        else:
            nb_recv = len(self.computer.input) // 3

        self.computer.resume()

        nb_to_send = len(self.computer.output) // 3

        while len(self.computer.output) != 0:
            dest_id = self.computer.output.popleft()
            X = self.computer.output.popleft()
            Y = self.computer.output.popleft()

            if part == 1 and dest_id == 255 and not answer:
                # if we exit here, part2 can't be chained
                answer = Y

            self.send_to(dest_id, X, Y)

        if nb_recv == 0 and nb_to_send == 0:
            self.idle.add(self.id)
        else:
            self.idle.discard(self.id)

        return answer

    def send_to(self, dest_id, X, Y):
        # Second arg is weird and mysterious: value seems to be ignored by the Intcode software.
        # Moreover, it doesn't work if I send only the 3 expected values (id,X,Y).
        # Found by accident during a code refactoring, that obviously did not work before.
        # Since then, I have reread the puzzle statement several times
        # and I still don't understand.
        self.network[dest_id].computer.input.extend((self.id, 0, X, Y))


nodes = [Node(id) for id in range(50)]
nodes.append(Node(255))  # add the NAT device

for node in itertools.cycle(nodes):
    Y = node.sched(1)
    if Y:
        print(Y)
        break

for node in itertools.cycle(nodes):
    Y = node.sched(2)
    if Y:
        print(Y)
        break
